src/astarte/handler.rs

```diff
@@ -203,81 +222,129 @@ impl DevicePublisher {
+            .await?;
 
-        let mut astarte_interfaces: Vec<Interface> = vec![];
+        let added_interfaces = added_names
+            .iter()
+            .filter_map(|i| node.introspection.get(i).cloned())
+            .collect();
 
-        for interface in astarte_node.introspection.iter() {
-            let astarte_interface: Interface = interface.clone().try_into()?;
-            self.client.add_interface(astarte_interface.clone()).await?;
+        let (sender, receiver) = channel(32);
 
-            astarte_interfaces.push(astarte_interface);
-        }
+        self.send_server_props(node, &sender).await?;
 
         self.subscribers.write().await.insert(
-            astarte_node.id,
+            node.id,
             Subscriber {
-                introspection: astarte_interfaces,
-                sender: tx,
+                introspection,
+                sender,
             },
         );
 
-        Ok(rx)
+        Ok(Subscription {
+            added_interfaces,
+            receiver,
+        })
     }
 
     /// Unsubscribe an existing Node and its introspection from Astarte Message Hub.
     ///
-    /// All the interfaces in this node introspection that are not in the introspection of any other node will be removed.
-    async fn unsubscribe(&self, astarte_node: &AstarteNode) -> Result<(), AstarteMessageHubError> {
-        let interfaces_to_remove = {
-            let subscribers_guard = self.subscribers.read().await;
-            astarte_node
+    /// All the interfaces in this node introspection that are not in the introspection of any other
+    /// node will be removed and the names will be returned.
+    async fn unsubscribe(&self, id: &Uuid) -> Result<Vec<String>, AstarteMessageHubError> {
+        let mut subscribers = self.subscribers.write().await;
+
+        let removed = {
+            let (id, node) = subscribers
+                .get_key_value(id)
+                .ok_or_else(|| AstarteMessageHubError::NodeId(*id))?;
+
+            let to_remove = node
                 .introspection
                 .iter()
-                .filter_map(|interface| interface.clone().try_into().ok())
-                .filter(|interface| {
-                    subscribers_guard
+                .filter(|&interface| {
+                    !subscribers
                         .iter()
-                        .filter(|(id, _)| astarte_node.id.ne(id))
-                        .find_map(|(_, subscriber)| {
-                            subscriber.introspection.contains(interface).then_some(())
+                        // Check if any other subscriber use the same interface
+                        .any(|(s_id, s_node)| {
+                            s_id != id && s_node.introspection.contains(interface)
                         })
-                        .is_none()
                 })
-                .collect::<Vec<astarte_device_sdk::Interface>>()
+                .collect_vec();
```

Comment:
could you avoid collecting the interfaces and directly use the iterator in the for loop?

Reply:
Indeed you can